doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md

```diff
@@ -0,0 +1,447 @@
+*/
+int my_set_handler(int argc, char **argv, char *val)
+{
+    if (argc) {
+        if (!strcmp("is_enabled", argv[0])) {
+            /* Set the value of `is_enabled` from `val` */
+            memcpy(&is_enabled, val, sizeof(is_enabled));
+        }
+        else if (!strcmp("threshold", argv[0])) {
+            /* Validate threshold */
+            if(atoi(val) > MAX_THRESHOLD)
+                return -EINVAL;
+
+            /* Set the value of `theshold` from `val` */
+            memcpy(&threshold, val, sizeof(threshold));
+        }
+    }
+    /* ... */
+    return 0;
+}
+
+/* Dummy implementation of `commit` handler.
+   This is intended to be called by the Registry when all configurations have
+   to be applied. Because of this, it's possible to implement transactions or
+   protect against faulty combinations of configs, race conditions, etc.
+*/
+int my_commit_handler(int argc, char **argv, char *val)
+{
+    /* Do something if the module is enable */
+    if(is_enabled) {
+        trigger_something();
+    }
+    
+    /* As stated before, the application crashes if `set_threashold` is called when is_enabled is false.
+    We protect it here */
+    if(is_enabled) {
+        /* We can safely set the theshold without crashing the app */
+        set_threshold(threshold);
+    }
+}
+
+/* Dummy implementation of `export` handler.
+   This handler has to call `export_func` for one or all parameters.
+   There can be different behaviors depending on the export function (e.g printing all configs
+   to STDOUT, save them in a non-volatile storage device, etc)
+*/
+void my_export_handler(int (*export_func)(const char *name, char *val), int argc,
+                   char **argv)
+{
+    /* argc  and argv can be used to export only one parameter */
+    (void)argv;
+    (void)argc;
+    char buf[INT_STRING_SIZE];
+
+    /* We export every parameter with the export function */
+    
+    /* Prepare `buf` to contain is_enabled in a string representation */
+    /* ... */
+    export_func("my_handler/is_enabled", buf);
+    /* Prepare `buf` to contain threshold in a string representation */
+    /* ... */
+    export_func("my_handler/threshold", buf);
+}
+```
+
+### Conceptual example of Storage Facility
+```c
+/*
+This conceptual example shows the implementation of a dummy Storage Facility.
+The storage device is a dummy_storage array (represented by the 
+`dummy_store_storage_t` struct).
+*/
+typedef struct {
+    char name[DUMMY_MAX_NAME_LEN];
+    char val[DUMMY_MAX_VAL_LEN];
+} dummy_store_storage_t;
+
+
+/* This will be our "storage device" */
+dummy_store_storage_t dummy_store[DUMMY_STORE_CAPACITY];
+
+static int registry_dummy_load(registry_store_t *store, load_cb_t cb,
+                               void *cb_arg);
+static int registry_dummy_store(registry_store_t *store, const char *name,
+const char *value);
+
+/* Storage Facility descriptor to be registered in the RIOT Registry */
+static registry_store_itf_t interface = {
```

Comment:
the example is just to show how the RH and SF might look. In practice, there might be a my_module_register() function in charge of the registration.
So, np with static IMO

Reply:
Ah right, it's illustrated [here](https://github.com/RIOT-OS/RIOT/blob/bf6b66e94acd8391f75610170eb0792436a0e951/doc/memos/rdm-draft-alamos-lanzieri-runtime-configuration-architecture.md#registry-initialization) as `registry_src_register(interface)`. Is it worth making the diagram indicate that the register function maybe takes a callback in that case?

Reply:
I think this is really implementation specific and not in the scope of the RDM

Reply:
Is it worth stating that? Otherwise the implementer might take the code and diagram as guidance and see that there's a discrepancy, and be left wondering which to implement

Reply:
> Otherwise the implementer might take the code and diagram as guidance and see that there's a discrepancy, and be left wondering which to implement

Honestly I don't see why we should state a very specific implementation detail in the RDM.
Static descriptors can indeed be registered in C. We do it all the time (e.g [netreg entries](https://github.com/RIOT-OS/RIOT/blob/68dc5b0d6e0422df9c3653c2cb8021fc35974aae/examples/gnrc_networking/udp.c#L128) )

Reply:
Ah ok, thanks for giving that netreg example, I was imagining some other startup code outside the module calling `registry_src_register`. Thanks for clarifying

Reply:
@danpetry as you mention, it says 'conceptual examples', the main use of these is to give a better idea of how the usage flow would be (once implemented). It is not expected to be used directly (the final API may not even be the same), for that purpose examples will be provided with the implementation once this RDM is approved. Given it will be in C I don't think representing it in pseudocode is needed IMHO.

Reply:
Maybe some clear text above saying that this must not be taken as a definition of the code? (NB that people will use this code anyway. At the very least you'll be leading people towards a certain implementation)